### PR TITLE
Use HMAC one-shot where appropriate

### PR DESCRIPTION
**Pull request by vcsjones**
Update to use the HMAC one shots where appropriate.

Most other places that use HMAC either A: Require appending (So `IncrementalHash` is the natural choice) or B: target downlevel frameworks.

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixExportProvider.cs [hunk 1]
```diff
@@ -475,44 +475,35 @@ private static ArraySegment<byte> EncodeKeys(AsnWriter tmpWriter, SafeBagAsn[] k
...
+            int bytesWritten = HMACSHA1.HashData(macKey, encodedAuthSafe.Span, macSpan);
 
-                CryptographicOperations.ZeroMemory(macKeySpan);
+            if (bytesWritten != MacSize)
+            {
+                Debug.Fail($"HMACSHA1.HashData wrote {bytesWritten} of {MacSize} bytes");
+                throw new CryptographicException();
             }
 
+            CryptographicOperations.ZeroMemory(macKey);
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixExportProvider.cs:505]: This is zeroing stack data now, but it's key material, so it seemed worth keeping. No strong feelings on this though.

**Review by krwq — APPROVED** [review]
Looks good to me. Do we have any sense of how much memory saving are we getting from this?

**Comment by vcsjones** [comment]
> Do we have any sense of how much memory saving are we getting from this?

The one-shots do not allocate for span in/span out usage (with the exception of Windows 7 and 8 allocating a handle). `IncrementalHash` has a number of allocations, including itself, `HMACCommon`, and various safe handles used by the PAL. I can put together benchmarks if it is desirable. But, since this improvement is not a case of "is this code gymnastics worth the allocation benefit?" I didn't do it initially.

**Comment by krwq** [comment]
I'll leave for @bartonjs and @GrabYourPitchforks to review since it's their area but I do not think the benchmark is required at this point.

**Review by bartonjs — APPROVED** [review]

**Comment by bartonjs** [comment]
Thanks, @vcsjones :smile:

**Pull request merged**